accepted/2020/minimum-os-version/minimum-os-version.md

```diff
@@ -0,0 +1,246 @@
+
+## OS Minimum Version
+
+The OS Minimum Version is the minimum version of the OS that an app or library supports running on. If it is not specified, it should default to the OS API Version. If an app or library author specifies an OS Minimum Version, they are declaring that they are taking responsibility for ensuring that the app can run on that version and higher by using runtime checks to guard calls to potentially unavailable APIs. This allows apps and libraries to progressively “light up” features on newer OS versions while continuing to run downlevel.
+
+We will make this concept first class and propagate the value throughout the build system and NuGet consistently, which will enable us to build analyzers and other experiences to help developers follow this pattern. This tooling will be detailed later in this document.
+
+# API Annotations
+
+To enable tooling to provide a good experience, the reference assemblies must be annotated with information about the availability of the APIs, as is current done for Xamarin Apple platforms.
+
+The following attributes are proposed for the `System.Runtime.Versioning` namespace. They are based on the Xamarin attributes, but they do not have the `Deprecated` availability kind (it’s redundant), they rename the `Unavailable` availability kind to `Removed` (for consistency), and they do not have the ability to differentiate availability based on architecture (reduction of scope).
+
+```csharp
+[AttributeUsage(AttributeTargets.All, AllowMultiple=true)]
+public abstract class PlatformAvailabilityAttribute
+{
+    internal PlatformAvailabilityAttribute(
+    string platformIdentifier,
+    int majorVersion, int minorVersion, int servicingVersion, int bindingRevision
+    string message
+    ) {}
+    public AvailabilityKind AvailabilityKind { get; }
+    public string Message { get; }
+    public string PlatformIdentifier { get; }
+    public Version OSVersion { get; }
+    public int BindingRevision { get; }
+}
+
+public enum AvailabilityKind
+{
+    Introduced,
+    Obsoleted,
+    Removed
+}
+```
+
+The platform string is expected to have the same values as used in the `TargetPlatformIdentifier` portion of platform-specific TFMs. Having it in the availability attributes may seem redundant as it will also be burned into assembly attributes for platform-specific reference assemblies. However, explicitly specifying the attribute is useful for platform-agnostic `net5.0` reference assemblies that have methods that are only available on some platforms e.g. due to having platform-specific implementations via bait-and-switch. This may also improve source readability when multi-targeting as it will remove the need to use `#if` around the annotations.
```

Comment:
@terrajobst is the thinking to apply such attributes to core libraries as well? For example when OpenSSL requires specific underlying version, or security APIs relying on specific PAL ios/macOS version?

Reply:
Even for platform-agnostic assemblies using platform-specific implementations should match the public surface APIs (aka the ref assembly should be one and only one for any number of platform specific implementations!)

But that need not be the case for platform-specific assemblies that might have different public APIs for each platform.

Reply:
Yes, I would assume we would add these to `net5.0` TFM.